rascaline/src/calculators/lode/spherical_expansion.rs

```diff
@@ -578,8 +591,25 @@ impl CalculatorBase for LodeSphericalExpansion {
 
         let mut descriptors_by_system = split_tensor_map_by_system(descriptor, systems.len());
 
+        // pick either parallelization over systems (if we have a lot of
+        // systems) or parallelization over samples.
+        //
+        // The parallelization over samples has a slightly higher overhead, so
+        // if we can avoid it we try to!
+        //
+        // Trying to use both at the same time would require additional
+        // synchronisation to ensure the right threads access per-system
+        // pre-computed data (i.e. anything that depends on k-vectors).
+        let (parallel_systems, parallel_samples) = if systems.len() > rayon::current_num_threads() {
+            (true, false)
+        } else {
+            (false, true)
+        };
+        let n_systems = systems.len();
+
```

Comment:
I think this is something we should out at a more general location in the code, so that all calculators can use this in the future. Maybe in `CalculatorBase`?

Reply:
I don't agree: the optimal strategy might depend on the algorithms, so this should be decided close to where the algorithms are executed.